satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,390 @@
+			if err != nil {
+				return Error.Wrap(err)
+			}
+
+		case *pb.StorageNodeMessage_Failed:
+			err = endpoint.handleFailed(ctx, &pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+		default:
+			return Error.New("unknown storage node message: %v", m)
+		}
+	}
+
+	if err := group.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (endpoint *Endpoint) sendPiecesToTransfer(ctx context.Context, stream processStream, nodeID storj.NodeID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	endpoint.log.Debug("sending pieces to transfer.", zap.String("nodeID", nodeID.String()))
+
+	return nil
+}
+
+func (endpoint *Endpoint) handleSucceeded(ctx context.Context, pending *sync.Map, nodeID storj.NodeID, message *pb.StorageNodeMessage_Succeeded) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	if message.Succeeded.GetAddressedOrderLimit() == nil {
+		return Error.New("Addressed order limit cannot be nil.")
+	}
+	if message.Succeeded.GetOriginalPieceHash() == nil {
+		return Error.New("Original piece hash cannot be nil.")
+	}
+	endpoint.log.Debug("transfer succeeded.", zap.String("piece ID", message.Succeeded.AddressedOrderLimit.Limit.PieceId.String()))
+
+	// TODO validation
+
+	transfer, ok := pending.Load(message.Succeeded.OriginalPieceHash.PieceId)
+	if !ok {
+		endpoint.log.Debug("could not find transfer message in pending queue. skipping .", zap.String("piece ID", message.Succeeded.AddressedOrderLimit.Limit.PieceId.String()))
+	}
+
+	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, nodeID, transfer.(pendingTransfer).path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	var failed int64
+	if transferQueueItem.FailedCount != nil && *transferQueueItem.FailedCount > 0 {
+		failed = -1
+	}
+
+	err = endpoint.db.IncrementProgress(ctx, nodeID, transfer.(pendingTransfer).pieceSize, 1, failed)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	err = endpoint.db.DeleteTransferQueueItem(ctx, nodeID, transfer.(pendingTransfer).path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	pending.Delete(message.Succeeded.GetAddressedOrderLimit().GetLimit().PieceId)
+
+	return nil
+}
+
+func (endpoint *Endpoint) handleFailed(ctx context.Context, pending *sync.Map, nodeID storj.NodeID, message *pb.StorageNodeMessage_Failed) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	endpoint.log.Warn("transfer failed.", zap.String("piece ID", message.Failed.PieceId.String()), zap.String("transfer error", message.Failed.GetError().String()))
+	pieceID := message.Failed.PieceId
+	transfer, ok := pending.Load(pieceID)
+	if !ok {
+		endpoint.log.Debug("could not find transfer message in pending queue. skipping .", zap.String("piece ID", pieceID.String()))
+	}
+	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, nodeID, transfer.(pendingTransfer).path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+	now := time.Now().UTC()
+	failedCount := 1
+	if transferQueueItem.FailedCount != nil {
```

Comment:
This logic would be simpler if `FailedCount` was a normal int rather than a pointer.

Reply:
Commented above on this